design/ratelimit-design-new.md

```diff
@@ -0,0 +1,397 @@
+| Local | [envoy.filters.network.local_ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/local_rate_limit_filter#config-network-filters-local-rate-limit) | [envoy.filters.http.local_ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter#config-http-filters-local-rate-limit) |
+| Global | [envoy.filters.network.ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/rate_limit_filter#config-network-filters-rate-limit) | [envoy.filters.http.ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter#config-http-filters-rate-limit) |
+
+This document focuses on the HTTP filters, as it's assumed that users will want L7 control over rate limiting.
+
+## Goals
+- Support Envoy's L7 **local** rate limiting filter.
+- Support Envoy's L7 **global** rate limiting filter, with a "bring your own Rate Limit Service (RLS)" model.
+
+## Non Goals
+- L4 rate limiting support (local or global).
+- Tight integration with a particular RLS.
+- Supporting more than one global RLS per Contour instance (since all non-TLS virtual hosts share an HTTP connection manager, this is not possible).
+
+## High-Level Design
+
+A new type, `RateLimitPolicy`, will be defined as part of the `HTTPProxy` API.
+A `RateLimitPolicy` can be defined for either a virtual host or a route.
+The `RateLimitPolicy` defines parameters for local and/or global rate limiting.
+
+
+For local rate limiting, the user defines the rate limit itself, via [token bucket](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto#envoy-v3-api-field-extensions-filters-http-local-ratelimit-v3-localratelimit-token-bucket) settings.
+
+
+For global rate limiting, the user defines the descriptors to be generated and sent to the external RLS.
+Descriptors contain entries including things like: the client IP, the value of a particular header, the destination cluster, etc.
+The external RLS makes the rate limit decision based on the descriptors, and returns either a 200 or a 429 to Envoy.
+The operator of the external RLS must configure it with actual rate limits for different descriptors.
+Each RLS implementation may have its own configuration format and mechanism for defining rate limits for descriptors, so Contour cannot provide a generic API for defining these.
+
+
+For global rate limiting, the details of the external RLS are configured via an `ExtensionService` and in the Contour config file.
+Note that the main reason this is defined globally for Contour in the config file, rather than per-virtual host, is that there's a rate limit filter per HTTP connection manager, and all non-TLS virtual hosts share a single HTTP connection manager, so they necessarily must share the same rate limit filter config.
+If no global RLS is defined, no global rate limiting will occur.
```

Comment:
Perhaps what would be useful here is a paragraph positioning global and local rate limiting. From one point of view you could see rate limiting as an abstract API that could be backed by either a global or a local rate limiter. However, I think the implication here is that local and global are solving different problems. Making that explicit (briefly) here would help clarify the API decisions and later (more detailed) help users approach the topic with the right mental model.